# Zoekmachines

## Notebook made by

__Name__|Yavuz| en vrienden

__Student id__ : 

## Toelichting

* De meeste opgaven worden automatisch nagekeken. Bij vrijwel alle opdrachten staan er een paar zichtbare tests onder de opdracht, dit is voornamelijk om te zorgen dat je de juiste type output geeft. De andere tests worden na inleveren toegevoegd aan die cell.

## Voor het inleveren!

* Pas niet de cellen aan, vooral niet die je niet kunt editen. Dit levert problemen op bij nakijken. Twijfel je of je per ongeluk iets hebt gewijzigd, kopieer dan bij inleveren je antwoorden naar een nieuw bestand, zodat het niet fout kan gaan.

* Zorg dat de code goed runt van boven naar beneden, verifieer dat door boven in Kernel -> Restart & Run All uit te voeren

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Zoekmachines" data-toc-modified-id="Zoekmachines-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Zoekmachines</a></span><ul class="toc-item"><li><span><a href="#Notebook-made-by" data-toc-modified-id="Notebook-made-by-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Notebook made by</a></span></li><li><span><a href="#Toelichting" data-toc-modified-id="Toelichting-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Toelichting</a></span></li><li><span><a href="#Voor-het-inleveren!" data-toc-modified-id="Voor-het-inleveren!-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Voor het inleveren!</a></span></li></ul></li><li><span><a href="#Week-5" data-toc-modified-id="Week-5-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Week 5</a></span></li><li><span><a href="#Evaluation-of-Document-Classification-(4pt)" data-toc-modified-id="Evaluation-of-Document-Classification-(4pt)-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Evaluation of Document Classification (4pt)</a></span></li><li><span><a href="#Relevance-feedback" data-toc-modified-id="Relevance-feedback-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Relevance feedback</a></span></li><li><span><a href="#Book-excercises-(Chapter-12)" data-toc-modified-id="Book-excercises-(Chapter-12)-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Book excercises (Chapter 12)</a></span><ul class="toc-item"><li><span><a href="#Exercise-12.3" data-toc-modified-id="Exercise-12.3-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Exercise 12.3</a></span></li><li><span><a href="#Exercise-12.4" data-toc-modified-id="Exercise-12.4-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Exercise 12.4</a></span></li><li><span><a href="#Exercise-12.5" data-toc-modified-id="Exercise-12.5-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Exercise 12.5</a></span></li><li><span><a href="#Exercise-12.6" data-toc-modified-id="Exercise-12.6-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Exercise 12.6</a></span></li><li><span><a href="#Exercise-12.7" data-toc-modified-id="Exercise-12.7-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>Exercise 12.7</a></span></li><li><span><a href="#Exercise-12.8" data-toc-modified-id="Exercise-12.8-5.6"><span class="toc-item-num">5.6&nbsp;&nbsp;</span>Exercise 12.8</a></span></li><li><span><a href="#Exercise-12.9" data-toc-modified-id="Exercise-12.9-5.7"><span class="toc-item-num">5.7&nbsp;&nbsp;</span>Exercise 12.9</a></span></li></ul></li></ul></div>

# Week 5

In [1]:
from nose.tools import assert_equal, assert_almost_equal
import numpy as np

# Evaluation of Document Classification (4pt)


Stel je hebt een Naive Bayes Document classifier getrained voor 1 klasse $C$. Stel je hebt een test-collectie van 400 documenten. 

Beantwoord de volgende vragen in de cellen hieronder.

1. De classifier heeft elk document in de collectie geklassificeerd. Wat betekent het als je een precisie hebt van 60%? Hoeveel documenten zijn dan onterecht als behorende tot klasse $C$ geklassificeerd?
2. Stel dat 300 van de test documenten in klasse $C$ vallen? Welke precisie baseline hoort bij deze test-collectie? Welke recall baseline? 
3. Stel je classifier heeft een recall van 50%. Stel 100 van de 400 docs zitten in klasse $C$.  De classifier hoeft niet elk document te classificeren.   Geef exact   aan wat de precisie kan zijn voor deze classifier. Geef dus de onder en bovengrens voor de precisie. 
4. Stel 100 van de 400 docs zitten in klasse $C$. Welke baseline heb je voor _accurracy_?

Met een classificatie baseline bedoelen we een classifier die helemaal niet naar de input kijkt maar alleen slim kiest op basis van de verdeling van de documenten over de klasse $C$.
Denk bijvoorbeeld aan de vraag naar het geslacht van een willekeurige KI student. Wat zou je kiezen als je niks van die student weet? Wat is de recall en de precisie voor het geslacht "jongen" dan?

In [3]:
# deel vraag 1
onterecht=160 
onterecht



160

In [4]:
# deel vraag 2

precisie_baseline= 300/(300+100)  
recall_baseline= 300/300



In [5]:
# deel vraag 3

ondergrens= 50/350  
bovengrens= 50/50



In [6]:
# deel vraag 4

accuracy_baseline= 100/400 
 
    


In [7]:
assert 0<=onterecht<= 400



In [8]:
assert 0<=precisie_baseline<=1
assert 0<=recall_baseline<= 1


In [9]:
assert 0<=ondergrens<=1
assert 0<=bovengrens<= 1

In [10]:
assert 0<=accuracy_baseline<=1


# Relevance feedback

In deze vraag gaan we de SMART versie van Rochio relevance feedback (vergelijking 9.3 in het boek) implementeren in Python. We gaan uit van een collectie met een vocabulair van 100 termen. Elke query en elk document is gerepresenteerd als een vector van lengte 100 gevuld met getallen tussen 0 en 1. 

Maak de functies voor `centroid` en `modified_query` af. Alle data zit in numpy arrays, dus je kunt er heel makkelijk mee werken op een manier die je kent uit linaire algebra. 

De data bestaat uit 
* een query van 2 woorden
* 8 relevante documenten en 3 niet relevante documenten

Hint: beide functies kan je zonder for loops schrijven, zelfs comprehensies zijn niet nodig.

In [11]:
# data
q=np.zeros(100)
q[8],q[47]= .4, .111
Relevant_docs=np.array([np.random.rand(100) for _ in range(8)])
NonRelevant_docs=np.array([np.random.rand(100) for _ in range(3)])

print(q)

[0.    0.    0.    0.    0.    0.    0.    0.    0.4   0.    0.    0.
 0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.
 0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.
 0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.111
 0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.
 0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.
 0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.
 0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.    0.
 0.    0.    0.    0.   ]


In [12]:
def centroid(D):
    
    return np.sum(D, axis = 0)/len(D)

# test
print(centroid(NonRelevant_docs ))

def modified_query(query,relevantdocs,nonrelevantdocs,alpha=1,beta=.75,gamma=.25):
    
    return alpha * query + beta * centroid(relevantdocs) - gamma * centroid(nonrelevantdocs)

# test

modified_query(q,Relevant_docs,NonRelevant_docs)



[0.49256593 0.377335   0.75264358 0.52698157 0.43630413 0.50001595
 0.57982357 0.61417493 0.31582894 0.19965316 0.63600346 0.7564291
 0.28266996 0.70999573 0.58286854 0.37276832 0.52951547 0.48395082
 0.3594933  0.51885877 0.48671166 0.68216535 0.63971676 0.56495293
 0.54086738 0.44056973 0.2750561  0.64358084 0.37243697 0.43705967
 0.53165023 0.43832933 0.09116858 0.57105882 0.55825556 0.32368387
 0.42856189 0.73452655 0.46638674 0.49571169 0.47878846 0.66905335
 0.26898043 0.34370542 0.31340432 0.58580012 0.46174647 0.64575032
 0.53200323 0.43779763 0.65550502 0.405358   0.43909324 0.27116357
 0.70132172 0.42270077 0.52967565 0.44167516 0.61788067 0.45473587
 0.67830547 0.57171111 0.23042287 0.35674676 0.50348467 0.77225952
 0.55169418 0.85957393 0.60912212 0.35598885 0.35023798 0.26101275
 0.27765988 0.5537294  0.48266865 0.2828379  0.58711163 0.77617705
 0.26734608 0.59538957 0.65250128 0.25859651 0.21226681 0.48185703
 0.37815282 0.35160134 0.41067851 0.37652129 0.45158356 0.78628

array([0.33767342, 0.22043823, 0.22818909, 0.397968  , 0.2382666 ,
       0.27006696, 0.18524984, 0.24736252, 0.51183699, 0.44115203,
       0.25924854, 0.32941842, 0.28418756, 0.16061271, 0.34760316,
       0.26421935, 0.27947976, 0.24504866, 0.39641541, 0.15203864,
       0.31008044, 0.13152647, 0.22910412, 0.1338627 , 0.10457301,
       0.29529124, 0.37308769, 0.16931539, 0.24574673, 0.36409032,
       0.11059464, 0.17022032, 0.33883787, 0.27156946, 0.06922223,
       0.38338441, 0.23139651, 0.09186728, 0.18624959, 0.33247564,
       0.25934463, 0.23283788, 0.25253798, 0.2959328 , 0.36784051,
       0.29742892, 0.23658498, 0.2998856 , 0.23113668, 0.27549312,
       0.27299636, 0.37706551, 0.29568375, 0.43492799, 0.12528171,
       0.33681815, 0.25442946, 0.3272988 , 0.2583233 , 0.14657316,
       0.17799218, 0.30227917, 0.31950097, 0.32972101, 0.3656171 ,
       0.13850843, 0.21558148, 0.12926875, 0.34214952, 0.20986672,
       0.30039813, 0.3624495 , 0.29847567, 0.32270069, 0.23080

In [13]:
assert isinstance(modified_query(q,Relevant_docs,NonRelevant_docs), np.ndarray)


# Book excercises (Chapter 12)
In this section you will make the exercises from the book, which is freely available online as a PDF at [nlp.stanford.edu/IR-book/](http://nlp.stanford.edu/IR-book/).

## Exercise 12.3
What  is  the  likelihood  ratio  of  the  document  according  to M1 and M2 in  Example 12.2?

In [14]:
likelihood_ratio = float(np.log((0.01*0.03*0.04*0.01*0.02*0.04*0.005)/(0.0002*0.03*0.04*0.0001*0.04*0.04*0.01)))

likelihood_ratio

7.1308988302963465

In [15]:
assert type(likelihood_ratio) in [int, float]

## Exercise 12.4
No explicit STOP probability appeared in Example 12.2. Assuming that the STOP probability of each model is 0.1, does this change the likelihood ratio of a document according to the two models?

In [16]:
likelihood_with_stop = float(np.log((0.01*0.03*0.04*0.01*0.02*0.04*0.005*0.1)/(0.0002*0.03*0.04*0.0001*0.04*0.04*0.01*0.1)))

likelihood_with_stop

7.1308988302963465

In [17]:
assert type(likelihood_with_stop) in [int, float]

## Exercise 12.5
How might a language model be used in a spelling correction system? In particular, consider the case of context-sensitive spelling correction, and correcting incorrect usages of words, such as _their_ in _Are you their?_(See Section3.5(page65) for pointers to some literature on this topic.)

A language model can be used in a spelling correction system by viewing the misspelled queries as probabilistic corruptions of a correct query. Using a probabilistic model, the misspellings can be corrected by taking the most probable correct query that is the closest in the given context.

## Exercise 12.6
Consider making a language model from the following training text:
    the martian has landed on the latin pop sensation ricky martin
    
a. Under a MLE-estimated unigram probability model, what are P(the) and P(martian)?

b. Under a MLE-estimated bigram model, what are P(sensation|pop) and P(pop|the)?

In [18]:
text = ["the", "martian", "has", "landed", "on", "the", "latin", "pop", "sensation", "ricky", "martin"]
P_the =  2/11
P_martian = 1/11
P_sensation_pop = 1/1
P_pop_the = 0/2

P_the, P_martian, P_sensation_pop, P_pop_the
len(text),text.count("the"), text.count("martian")

(11, 2, 1)

In [19]:
assert type(P_the) in [int, float]

In [20]:
assert type(P_martian) in [int, float]

In [21]:
assert type(P_sensation_pop) in [int, float]

In [22]:
assert type(P_pop_the) in [int, float]

## Exercise 12.7
Suppose we have  a collection that consists of the 4 documents given in the below table.

| docID | Document text                              |
|-------|--------------------------------------------|
| 1     | click go the shears boys click click click |
| 2     | click click                                |
| 3     | metal here                                 |
| 4     | metal shears click here                    |


Build a  query likelihood language  model for  this  document  collection. Assume  amixture model between the documents and the collection, with both weighted at 0.5. Maximum likelihood estimation (mle) is used to estimate both as unigram models. Work out the model probabilities of the queries `click`, `shears`, and hence `click shears` for each document, and use those probabilities to rank the documents returned by eachquery.

For each query give the probablities in a list, where the first value corresponds to Doc1, the second value to Doc2 etc.

In [23]:
P_click = [(4/8) * 0.5 + (7/16) * 0.5, (2/2) * 0.5 + (7/16) * 0.5, (0/2) * 0.5 + (7/16) * 0.5, (1/4) * 0.5 + (7/16) * 0.5]
P_shears = [(1/8) * 0.5 + (2/16) * 0.5, (0/2) * 0.5 + (2/16) * 0.5, (0/2) * 0.5 + (2/16) * 0.5, (1/4) * 0.5 + (2/16) * 0.5]
P_click_shears = [(((4/8) * 0.5 + (7/16) * 0.5) * ((1/8) * 0.5 + (2/16) * 0.5)), (((2/2) * 0.5 + (7/16) * 0.5) * ((0/2) * 0.5 + (2/16) * 0.5)), (((0/2) * 0.5 + (7/16) * 0.5) * ((0/2) * 0.5 + (2/16) * 0.5)), (((1/4) * 0.5 + (7/16) * 0.5) * ((1/4) * 0.5 + (2/16) * 0.5))]
# Final ranking 4 > 1 > 2 > 3
P_click, P_shears, P_click_shears

([0.46875, 0.71875, 0.21875, 0.34375],
 [0.125, 0.0625, 0.0625, 0.1875],
 [0.05859375, 0.044921875, 0.013671875, 0.064453125])

In [24]:
assert_equal(type(P_click), list)
assert_equal(len(P_click), 4)

In [25]:
assert_equal(type(P_shears), list)
assert_equal(len(P_shears), 4)

In [26]:
assert_equal(type(P_click_shears), list)
assert_equal(len(P_click_shears), 4)

## Exercise 12.8
Using the  calculations in Exercise 12.7 as inspiration or as examples where appropriate, write one sentence each  describing  the  treatment  that  the  model  in  Equation (12.10) gives to each of the following quantities. Include whether it is present in the model or not and whether the effect is raw or scaled.

a. Term frequency in a document

b. Collection frequency of a term

c. Document frequency of a term

d. Length normalization of a term

a) present: This is used to calculate the probability of the term appearing in the document. It is calculated by diving the term frequency by the total amount of words in the document.

b) present: This is used to calculate the probability of the term appearing in the collection. It is calculated by dividing the term frequency by the total amount of words in the collection.

c) not present

d) not present

## Exercise 12.9
In the mixture model approach to the query likelihood model (Equation (12.12)), the probability estimate of a term is based on the term frequency of a word in a document, and the collection frequency of the word.  Doing this certainly guarantees that each term of a query (in the vocabulary) has a non-zero chance of being generated by each document. But it has a more subtle but important effect of implementing a form of term weighting, related to what we saw in Chapter 6. Explain how this works. In particular, include  in  your answer a concrete numeric example showing  this  termweighting at work.

When a word often occurs in a collection and its documents, it loses its worth. Words that are more unique, have a higher weighting because they are more rare. The mixture model makes sure of this, by lowering the worth of words that have a high term frequency. This is because they don't have any meaningful addition to the query if every document is returned because they have this often-occuring word. 

An example of this is as the following: <br>
p(The) = 0.9<br>
p(Automobile) = 0.4<br>
Original value of 'The automobile': 0.9 * 0.4 = 0.36<br>
New value of 'The automobile' with weighting: (0.1 * 0.9) * ((1-0.1) * 0.4) = 0.0324